tests/backward-compat-recovery-no-password/src/test/groovy/org/apache/bookkeeper/tests/TestCompatRecoveryNoPassword.groovy

```diff
@@ -0,0 +1,248 @@
+
+        // Create a 4.1.0 client, will update /ledgers/LAYOUT
+        def v410CL = MavenClassLoader.forBookKeeperVersion("4.1.0")
+        def v410Conf = v410CL.newInstance("org.apache.bookkeeper.conf.ClientConfiguration")
+        v410Conf.setZkServers(zookeeper).setLedgerManagerType("hierarchical")
+        def v410BK = v410CL.newInstance("org.apache.bookkeeper.client.BookKeeper", v410Conf)
+
+        // Start bookies
+        def bookieContainers = new ArrayList<>(DockerUtils.cubeIdsMatching("bookkeeper"))
+        Assert.assertTrue(bookieContainers.size() >= 3)
+        Assert.assertTrue(BookKeeperClusterUtils.startBookieWithVersion(
+                docker, bookieContainers.get(0), currentVersion))
+        Assert.assertTrue(BookKeeperClusterUtils.startBookieWithVersion(
+                docker, bookieContainers.get(1), currentVersion))
+
+        // recreate bk client so that it reads bookie list
+        v410BK.close()
+        v410BK = v410CL.newBookKeeper(zookeeper)
+
+        // Write a ledger
+        def ledger0 = v410BK.createLedger(2, 2,
+                                          v410CL.digestType("MAC"), passwdCorrect)
+        for (int i = 0; i < numEntries; i++) {
+            ledger0.addEntry("foobar".getBytes())
+        }
+        ledger0.close()
+        v410BK.close()
+
+        // start a new bookie, and kill one of the initial 2
+        def failedBookieId = new BookieSocketAddress(
+            DockerUtils.getContainerIP(docker, bookieContainers.get(0)), 3181)
+        Assert.assertTrue(BookKeeperClusterUtils.stopBookie(
+                docker, bookieContainers.get(0)))
+        Assert.assertTrue(BookKeeperClusterUtils.startBookieWithVersion(
+                docker, bookieContainers.get(2), currentVersion))
+
+        def bkCur = new BookKeeper(zookeeper)
+
+        LedgerHandle lh = bkCur.openLedgerNoRecovery(
+            ledger0.getId(), BookKeeper.DigestType.MAC, passwdCorrect)
+        Assert.assertFalse("Should be entries missing",
+                           verifyFullyReplicated(bkCur, lh, numEntries))
+        lh.close()
+
+        ClientConfiguration adminConf = new ClientConfiguration()
+        adminConf.setZkServers(zookeeper)
+        adminConf.setBookieRecoveryDigestType(BookKeeper.DigestType.MAC)
+        adminConf.setBookieRecoveryPasswd(passwdBad)
+
+        def bka = new BookKeeperAdmin(adminConf)
+        try {
+            bka.recoverBookieData(failedBookieId)
+            Assert.fail("Shouldn't be able to recover with wrong password")
+        } catch (BKException bke) {
+            // correct behaviour
+        } finally {
+            bka.close();
+        }
+
+        adminConf.setBookieRecoveryDigestType(BookKeeper.DigestType.CRC32)
+        adminConf.setBookieRecoveryPasswd(passwdCorrect)
+
+        bka = new BookKeeperAdmin(adminConf)
+        try {
+            bka.recoverBookieData(failedBookieId)
+            Assert.fail("Shouldn't be able to recover with wrong digest")
```

Comment:
I don't know if we can use 'import static' in Groovy.

I wonder if we can have some flavour of checkstyle for Groovy

/cc @sijie

Reply:
I don't think we enforce any checkstyle about import static. We only disallow "import *". We are okay to use "Assert.fail" or "import static ..fail;" and use "fail". 

Also I don't know how checkstyle is applied to groovy. I think we need an issue to enforce checkstyle on groovy as well.

Reply:
Create #1043